AR-js-org/AR.js
# Webpack Build for AR.js

:information_source: _This is a **draft** PR. Its aim is to provide a forum to discuss the changes introduced and seek some help in order to make the PR merge-ready._

## What kind of change does this PR introduce?
Webpack build-system instead of "`cat` build-system" :wink: , module-based code, usage of [andypotato's
artoolkit5-js](https://github.com/andypotato/artoolkit5-js) 

## Related issues
- #21 
- #68 

## Summary
In order to comfortably integrate this great library in a modern web application (React, Angular, Ember.js, ...) it would be nice to have AR.js provided as a **ES6 module** (while keeping support for browser `<script>` usage). Refactoring the library to also use import/export internally, has the additional (and arguably most important) advantage of **improved maintainability**.  
In order to achieve the above, I refactored the library to use imports/exports instead of relying on globals and introduced a build configuration for [webpack](https://webpack.js.org/). Since the [jsartoolkit5](https://github.com/artoolkitx/jsartoolkit5) library is somewhat [flawed](https://github.com/artoolkitx/jsartoolkit5/issues/69) when it comes to ES6 support, this PR depends on the work done by @andypotato in [artoolkit5-js](https://github.com/andypotato/artoolkit5-js) to make the library ES6-compatible.  

I started off with getting the regular (non-nft) build to work and feel like I'm quite far, but got somewhat stuck now. When running the [basic](https://github.com/AR-js-org/AR.js/blob/da5ea830211c323a41ebd55a97b99f85add1cd40/aframe/examples/marker-based/basic.html) example, the initialization process goes well, markers are detected when in view (it runs through [updateWithModelViewMatrix](https://github.com/MikiDi/AR.js/blob/0a69d3696fade29845702c67558c8aa79ab0ef8a/three.js/src/threex/arjs-markercontrols.js#L107)), but the gltf **model ( :t-rex: ) doesn't show**. When debugging I discovered the [A-Frame inspector](https://github.com/aframevr/aframe-inspector) (I have little prior experience with aframe/three.js). The weirdest thing is that when you run the example with the A-Frame inspector, everything works as expected! :open_mouth:  This leads me to suspect that there's a minor bug in the aframe/three.js related part, which perhaps can be easily solved with some more background knowledge ... I can't seem to figure it out ...
Before continuing work to get the NFT build going (and on a longer term refactor out the separate builds and code duplication between nft/non-nft ), I'd like to get this fixed ... Is this something you could help out with maybe @kalwalt @nicolocarpignoli ?

## Does this PR introduce a breaking change?
No. Functionality-wise, the end-result of this PR should do exactly the same as the current master build.

## How can we test it?
Host the [basic](https://github.com/AR-js-org/AR.js/blob/da5ea830211c323a41ebd55a97b99f85add1cd40/aframe/examples/marker-based/basic.html) example and use the  `aframe-ar.js` build from the [PR branch](https://github.com/MikiDi/AR.js/tree/webpack/aframe/build).
Add [A-Frame inspector](https://github.com/aframevr/aframe-inspector#specifying-inspector-build) and press `<ctrl> + <alt> + i` to inject the inspector (make sure to press :play_or_pause_button: )  
In order to make changes and rebuild:
```
npm install
```
```
npm run build:dev
```

## Device used for tests, version of OS and version of Browser
Desktop (webcam), Fedora 30, Firefox 74.0
